nuxeo-core/nuxeo-core-io/src/main/java/org/nuxeo/ecm/core/io/avro/DocumentModelMapper.java

```diff
@@ -0,0 +1,98 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Contributors:
+ *     pierre
+ */
+package org.nuxeo.ecm.core.io.avro;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.avro.Schema;
+import org.apache.avro.Schema.Field;
+import org.apache.avro.Schema.Type;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericRecord;
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.api.impl.DocumentModelImpl;
+import org.nuxeo.ecm.core.api.model.Property;
+import org.nuxeo.runtime.RuntimeServiceException;
+import org.nuxeo.runtime.avro.AvroMapper;
+import org.nuxeo.runtime.avro.AvroService;
+
+/**
+ * @since 10.2
+ */
+public class DocumentModelMapper extends AvroMapper<DocumentModel, GenericRecord> {
+
+    private static final String ID = "uuid";
+
+    private static final String PATH = "path";
+
+    private static final String DOCUMENT_TYPE = "documentType";
+
+    private static final String DOCUMENT_MODEL = "documentModel";
+
+    public DocumentModelMapper(AvroService service) {
+        super(service);
+    }
+
+    @Override
+    public DocumentModel fromAvro(Schema schema, GenericRecord input) {
+        if (!DOCUMENT_MODEL.equals(getLogicalType(schema))) {
+            throw new RuntimeServiceException("Schema does not match DocumentModel");
+        }
+        String path = (String) input.get(PATH);
+        GenericRecord documentTypeRecord = (GenericRecord) input.get("primaryType");
+        DocumentModel doc = new DocumentModelImpl((String) null, path, documentTypeRecord.getSchema().getName());
+        for (Field schemaField : documentTypeRecord.getSchema().getFields()) {
+            GenericRecord schemaRecord = (GenericRecord) documentTypeRecord.get(schemaField.name());
+            List<Field> fields = schemaField.schema().getFields();
+            Map<String, Object> data = new HashMap<>(fields.size());
+            for (Field field : fields) {
+                Object value = service.fromAvro(field.schema(), Property.class, schemaRecord.get(field.name()));
+                data.put(service.decodeName(field.name()), value);
+            }
+            doc.setProperties(service.decodeName(schemaField.name()), data);
+        }
+        return doc;
+    }
+
+    @Override
+    public GenericRecord toAvro(Schema schema, DocumentModel input) {
+        String logicalType = getLogicalType(schema);
+        GenericRecord record = new GenericData.Record(schema);
+        if (DOCUMENT_MODEL.equals(logicalType)) {
+            record.put(PATH, input.getPathAsString());
+            record.put(ID, input.getId());
+        }
+        for (Field field : schema.getFields()) {
+            if (field.name().equals(ID) || field.name().equals(PATH)) {
```

Comment:
Is there something preventing us to treat ID and PATH the same way we did for others properties ?

I'm thinking about a schema containing `uuid` field.

Reply:
nop, but it should, it's the object of the next avro ticket around ecm properties